### PR TITLE
rework WriteBufferFromS3, add tests, add abortion

### PR DESCRIPTION
**Pull request by CheSema**
- introducing simple mock for S3 client
  - it tests in sync and async mode
  - injecting errors
- some tests for part splitting in multi part upload
- less copy operation inside WriteBufferFromS3 due to attaching memory to the iostream
- test for limiting part size for task:
  - https://github.com/ClickHouse/ClickHouse/issues/44271
  - buffer could produce parts >5G, would be generated 5G + 1m parts.
- simplify task synchronization

Other thing related to the topic, deserve to be as separate issues,
- limit active tasks count by memory size in allocated buffers
- DO NOT CALL finalize inside destructors (WriteIndirectBufferFromRemoteFS, WriteBufferFromFileDecorator) within no-throw scope. That breaks the contract with, for example, WriteBufferFromS3. A file is written to S3 only when finalize has finished successfully. Calling finalize in no-throw scope doesn't get any garanty about write status. 
- @vitlibar propose to use simpler methods set and position instead swapping for delegation implementation.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/IO/WriteBufferFromS3.cpp [hunk 1]
```diff
@@ -217,30 +242,27 @@ void WriteBufferFromS3::createMultipartUpload()
...
+    LOG_TRACE(log, "Writing part. Bucket: {}, Key: {}, Upload_id: {}, Size: {}", bucket, key, multipart_upload_id, size);
+
     if (TSA_SUPPRESS_WARNING_FOR_READ(part_tags).size() == S3_WARN_MAX_PARTS)
     {
         // Don't throw exception here by ourselves but leave the decision to take by S3 server.
         LOG_WARNING(log, "Maximum part number in S3 protocol has reached (too many parts). Server may not accept this whole upload.");
     }
 
+    chassert(size <= settings.max_upload_part_size);
+    chassert(size >= settings.min_upload_part_size || last_part);
```
Comment (KochetovNicolai) [inline src/IO/WriteBufferFromS3.cpp:264]: I think it's better to throw logical error here.
This check is cheap and better to detect in release build as well.

File: src/IO/WriteBufferFromS3.cpp [hunk 2]
```diff
@@ -94,41 +105,56 @@ WriteBufferFromS3::WriteBufferFromS3(
...
 
-    chassert(upload_part_size > 0);
-    if (!multipart_upload_id.empty() && last_part_size > upload_part_size)
+    if (!multipart_upload_id.empty())
     {
-        writePart();
-
-        allocateBuffer();
+        if (offset() == memory.size())
+            writePart();
```
Comment (KochetovNicolai) [inline src/IO/WriteBufferFromS3.cpp:123]: It looks like we expect here that `nextImpl` is called only when the buffer is full. I suspect that it is not always true (this is not an invariant that we check/support for buffers).

Comment (CheSema) [reply src/IO/WriteBufferFromS3.cpp:123]: It just has to be true for such implementation. If it is not than there is no way to implement one-copy buffer. As long as we need to control parts sizes we have to either do double-copy or demand that nextImpl is call when buffer is full (last part is ok to be lesser).

Comment (CheSema) [reply src/IO/WriteBufferFromS3.cpp:123]: But you are right in that thing: there is happens call next which is supposed to flush the buffer. And WriteBufferFromS3 has to flush it due to realisation WriteBuffer::next. WriteBuffer::next just rewind the buffer what ever what happens in successor. And this is happening indeed, I see "DB::Exception: Failed to complete multipart upload. No parts have uploaded." in CI.
I'm thinking what to do with that case.

Comment (CheSema) [reply src/IO/WriteBufferFromS3.cpp]: So, I find the way to address it.
I don't have much choice when next is called. I prefer to flush the buffer and make a new part with current partially filled buffer. That could lead to creating parts with sizes smaller than 5M.  If it happens in production enviroment S3 returns an error. I suppose that we do not do such thing. I added assert statement in order to check it with debug build. Any way, all builds would have an LOG_WARNING message for such case.

File: src/IO/WriteBufferFromS3.cpp (hunk 1 above)
Comment (CheSema) [reply src/IO/WriteBufferFromS3.cpp]: I reworked that part.

**Comment by serxa** [comment]
Note that there are now two places where multipart upload is used independently https://github.com/ClickHouse/ClickHouse/pull/45487

File: src/IO/WriteBufferFromS3.h [hunk 3]
```diff
@@ -52,75 +44,121 @@ class WriteBufferFromS3 final : public BufferWithOwnMemory<WriteBuffer>
...
+    bool multipart_upload_aborted = false;
+
+    /// Track that prefinalize() is called only once
+    bool is_prefinalized = false;
+
+    /// First fully filled buffer has to be delayed
+    /// There are two ways after:
+    /// First is to call prefinalize/finalize, which leads to single part upload
+    /// Second is to write more data, which leads to multi part upload
+    std::deque<PartData> detached_part_data;
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3.h:153]: If I understood correctly this PR can help to eliminate extra copying to `std::stringstream`, but it can't help with huge memory consumption because you have to keep the data written to `WriteBufferFromS3` anyway until aws sdk uses it.

File: src/IO/SwapHelper.h [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
-    {
-    public:
-        SwapHelper(BufferBase & b1_, BufferBase & b2_) : b1(b1_), b2(b2_) { b1.swap(b2); }
-        ~SwapHelper() { b1.swap(b2); }
 
-    private:
-        BufferBase & b1;
-        BufferBase & b2;
-    };
+class SwapHelper
```
Comment (vitlibar) [inline src/IO/SwapHelper.h:8]: This class looks strange.

File: src/IO/WriteBufferFromS3.h (hunk 3 above)
Comment (CheSema) [reply src/IO/WriteBufferFromS3.h:153]: Yes. It is right. To handle it we need bound unfinished tasks. That problem hasn't been addressed here.

File: src/IO/SwapHelper.h (hunk 4 above)
Comment (CheSema) [reply src/IO/SwapHelper.h:8]: I also do not like that, but what exactly could be better?

File: src/IO/WriteBufferFromFileDecorator.cpp [hunk 5]
```diff
@@ -31,11 +39,18 @@ WriteBufferFromFileDecorator::~WriteBufferFromFileDecorator()
...
         tryLogCurrentException(__PRETTY_FUNCTION__);
     }
+
+    swap(*impl);
 }
 
 void WriteBufferFromFileDecorator::sync()
 {
-    impl->sync();
+    next();
```
Comment (CheSema) [inline src/IO/WriteBufferFromFileDecorator.cpp:48]: That line costs me several days of debugging.

WriteIndirectBufferFromRemoteFS::count() is incorrect if impl->next called outside of WriteIndirectBufferFromRemoteFS::next(). That leads to the situation when file size is set to 0 in the file system for file format_version.txt.

File: src/IO/WriteBufferFromS3.h (hunk 3 above)
Comment (vitlibar) [reply src/IO/WriteBufferFromS3.h:153]: What are bound unfinished tasks? For me it doesn't look possible to solve the memory consumption issue in general for big `upload_part_size` without making an intermediate copy to a local disk.

File: src/IO/WriteBufferFromFileDecorator.cpp [hunk 6]
```diff
@@ -15,10 +16,17 @@ void WriteBufferFromFileDecorator::finalizeImpl()
...
+    /// WriteBufferFromFileDecorator.finalized is set as true
+    /// but impl->finalized is remain false
+    /// That leads to situation when the destructor of impl is called with impl->finalized equal false.
     if (!is_prefinalized)
         WriteBufferFromFileDecorator::preFinalize();
 
-    impl->finalize();
+    {
+        SwapHelper swap(*this, *impl);
+        impl->finalize();
```
Comment (vitlibar) [inline src/IO/WriteBufferFromFileDecorator.cpp:28]: Can you explain what is going on here? Why do we need this `SwapHelper` and why can't we just call `impl->finalize()`?

Comment (CheSema) [reply src/IO/WriteBufferFromFileDecorator.cpp:28]: > @CheSema Can you explain what is going on here? Why do we need this `SwapHelper` and why can't we just call `impl->finalize()`?


First of all, is it clear why SwapHelper is used in nextImpl call?
If it is, then the answer is short: impl has to see its working_buffer, not the working_buffer(nullptr, nullptr) and pos(nullptr). Otherwise impl ends up in a situation when working_buffer and pos were driven away from him.

Long read:
 WriteBufferFromFileDecorator is created with nulls in working_buffer and pos. It is an empty buffer. At the constructor WriteBufferFromFileDecorator extract  working_buffer and pos from impl with swap call. Due to which WriteBufferFromFileDecorator::write writes data into working_buffer from impl. This is important part. When the working_buffer is filled out, WriteBufferFromFileDecorator::next is called. Here WriteBufferFromFileDecorator::bytes is updated (also important var) and WriteBufferFromFileDecorator::nextImpl called. In this part WriteBufferFromFileDecorator::nextImpl delegates the call to the impl->next. But Impl needs to see its working_buffer and pos, for it SwapHelper is called. Here the bug is hidden couse next is called right before.  That call provide working_buffer from impl to the impl in right way and impl is able to consume the data.

The same things happens when WriteBufferFromFileDecorator delegates any call to Impl. And it is a BUG that WriteBufferFromFileDecorator missed swaps in some functions including destructor.

File: src/IO/WriteBufferFromS3.h (hunk 3 above)
Comment (CheSema) [reply src/IO/WriteBufferFromS3.h:153]: Well. the simpliest solution could be as the following:
- make one_buffer, push it to thread pool, make account bytes_in_fly += one_buffer.size()
- make other_buffer, push it to thread pool, make account bytes_in_fly += other_buffer.size()
- make third_buffer, if bytes_in_fly is too big, then hung and wait until some task is finished
- for example one is finished, we do bytes_in_fly -= one_buffer.size(), that unblock third_buffer buffer to proceed further
That I meant by "bound unfinished tasks" by it's sizes.

File: src/IO/WriteBufferFromFileDecorator.cpp (hunk 5 above)
Comment (CheSema) [reply src/IO/WriteBufferFromFileDecorator.cpp:48]: It is interesting fact, why old WriteBufferFromFileDecorator::sync didn't couse any fails in test?
Because old WriteBufferFromFileDecorator::sync did nothing at all. impl->sync has been called with nulls inside working_buffer and pos du to missing swaps. When I added SwapHelper, impl->sync has began do the work, as a result WriteBufferFromFileDecorator::bytes and Impl.bytes have diverged.

**Comment by CheSema** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/44869/ab7576197d27221daf3e5e5ef1e5f4749c6c3927/stateless_tests__asan__[1/4].html
Nice catch!
02241_filesystem_cache_on_write_operations

File tmp_mutation_6.txt missed in the cache du to incorrect WriteBufferFromFileDecorator::sync() implementation.
https://github.com/ClickHouse/ClickHouse/blob/ed27987646897615fe9a855a57ca591a01b81ca2/src/Storages/MergeTree/MergeTreeMutationEntry.cpp#L76
And of course, finalize has to be called! Otherwise the is no guarantee that file is written.

File: src/IO/WriteBufferFromS3.h (hunk 3 above)
Comment (vitlibar) [reply src/IO/WriteBufferFromS3.h:153]: AWS SDK uses seeking on passed `std::stream` so this way you'll never be able to deallocate any buffers.

File: src/IO/WriteBufferFromFileDecorator.cpp (hunk 6 above)
Comment (vitlibar) [reply src/IO/WriteBufferFromFileDecorator.cpp:27]: I don't understand why we do that swapping at all. I understood that the constructor of WriteBufferFromFileDecorator need to get `working_buffer` from `impl`. But why can't the constructor of `WriteBufferFromFileDecorator` just call `BufferBase::set()` and then in `nextImpl` set the position again? See for example how `ReadBufferWrapper` is implemented. Swapping is two-directional and it changes both sides and it's more difficult to understand. So I suppose if just `set()` can do the job it's better not use swapping.

Comment (CheSema) [reply src/IO/WriteBufferFromFileDecorator.cpp:27]: I wouldn't advocate swapping at all. Here I found the problem in swapping implementation and fixed it.
You suggestion how to implement delegation it in a different way seems good. We should try it to make things simpler. I do not with to introduce such changes right here, but I would be happy to try implement it in following PR.

File: src/IO/WriteBufferFromS3.h (hunk 3 above)
Comment (CheSema) [reply src/IO/WriteBufferFromS3.h:152]: > AWS SDK uses seeking on passed `std::stream` so this way you'll never be able to deallocate any buffers until the whole part (which is up to 5GB) is uploaded

Yes. that is correct.

Comment (CheSema) [reply src/IO/WriteBufferFromS3.h:152]: > What are the bound unfinished tasks? I doubt that it's possible to solve the memory consumption issue in general for big `upload_part_size` at all without making an intermediate copy to a local disk.

Interesting approach. That will work for sure. In such way we definetly should care about an extra memcpu :)

Comment (vitlibar) [reply src/IO/WriteBufferFromS3.h:152]: If you manage to get rid of seeking inside AWS SDK in order to calculate checksums, then it would be better to change AWS SDK a bit to teach it work directly with our WriteBuffer, without any intermediate buffering at all. I mean in this case `WriteBufferFromS3` could directly send ready blocks of data to network.

**Comment by CheSema** [comment]
01164_detach_attach_partition_race flaky https://github.com/ClickHouse/ClickHouse/issues/46109

File: src/IO/S3/Client.cpp [hunk 7]
```diff
@@ -178,70 +178,70 @@ Model::HeadObjectOutcome Client::HeadObject(const HeadObjectRequest & request) c
...
     request.overrideURI(std::move(*bucket_uri));
 
-    return Aws::S3::S3Client::HeadObject(request);
+    return HeadObject(static_cast<const Model::HeadObjectRequest&>(request));
 }
 
 Model::ListObjectsV2Outcome Client::ListObjectsV2(const ListObjectsV2Request & request) const
 {
-    return doRequest(request, [this](const Model::ListObjectsV2Request & req) { return Aws::S3::S3Client::ListObjectsV2(req); });
+    return doRequest(request, [this](const Model::ListObjectsV2Request & req) { return ListObjectsV2(req); });
```
Comment (vitlibar) [inline src/IO/S3/Client.cpp:186]: `doRequest` calls a passed function multiple times. It seems you're making this `Client::ListObjectsV2` recursive here, it doesn't look correct.

File: src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp [hunk 8]
```diff
@@ -160,7 +160,10 @@ void MergeTreeDeduplicationLog::rotate()
     existing_logs.emplace(current_log_number, log_description);
 
     if (current_writer)
+    {
+        current_writer->finalize();
         current_writer->sync();
```
Comment (vitlibar) [inline src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp:165]: Why are you calling `sync` after `finalize`? Why do you even need both of them? I mean mustn't `finalize()` include `sync()`?

File: src/IO/WriteBufferFromFileDecorator.cpp [hunk 9]
```diff
@@ -31,11 +38,18 @@ WriteBufferFromFileDecorator::~WriteBufferFromFileDecorator()
     {
         tryLogCurrentException(__PRETTY_FUNCTION__);
     }
+
+    swap(*impl);
```
Comment (vitlibar) [inline src/IO/WriteBufferFromFileDecorator.cpp:42]: What's the point of calling `swap` here? It's after `finalize()`, the buffer isn't going to be used anymore.

File: src/IO/WriteBufferFromS3.cpp [hunk 10]
```diff
@@ -72,120 +191,264 @@ WriteBufferFromS3::WriteBufferFromS3(
...
-    last_part_size += offset();
-    if (write_settings.remote_throttler)
-        write_settings.remote_throttler->add(offset(), ProfileEvents::RemoteWriteThrottlerBytes, ProfileEvents::RemoteWriteThrottlerSleepMicroseconds);
+    if (available() > 0)
+        return;
 
-    /// Data size exceeds singlepart upload threshold, need to use multipart upload.
-    if (multipart_upload_id.empty() && last_part_size > upload_settings.max_single_part_upload_size)
-        createMultipartUpload();
+    detached_part_data.push_back(detachBuffer());
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3.cpp:225]: The only thing you do with the result of `detachBuffer()` is appending it to the end of `detached_part_data`, so I think that appending can be made a part of `detachBuffer()`.

File: src/IO/S3/Client.cpp (hunk 7 above)
Comment (CheSema) [reply src/IO/S3/Client.cpp:186]: No. There is no recursion.
The functions
- Client::ListObjectsV2(const ListObjectsV2Request&) and 
- Aws::S3::S3Client::ListObjectsV2(const Model::ListObjectsV2Request &)
are different functions. They have different argument type. They are not overwrite each other. Aws::S3::S3Client::ListObjectsV2 is virtual, meanwhile Client::ListObjectsV2 is not.

Here I do a possibility to call ListObjectsV2(const Model::ListObjectsV2Request) by virtual call. This is needed to have an ability to override the virtual calls Aws::S3::S3Client::* for mocking client.

File: src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp (hunk 8 above)
Comment (CheSema) [reply src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp:165]: > Why are you calling `sync` after `finalize`? Why do you even need both of them? I mean mustn't `finalize()` include `sync()`?

Well. This is tricky.  sync is meaningless for S3. finalize is crucial  for S3.
That interface is written and broadly used with files. For files there are different priorities. With files exactly such approach is used. I was trying to make it consistent with other use cases. For files finalize do final operation in buffer to write the file, sync calls syscall for flush the filesystem buffers. And sync is called exactly after finish.

File: src/IO/WriteBufferFromFileDecorator.cpp (hunk 9 above)
Comment (CheSema) [reply src/IO/WriteBufferFromFileDecorator.cpp:42]: After finalize call the impl's destructor is called. It is crucial to call it with the right buffers. Not with nullptr's. WriteBufferFromFileDecorator calls swap at constructor. It is has to be call at destructor.

File: src/IO/WriteBufferFromS3.cpp (hunk 10 above)
Comment (CheSema) [reply src/IO/WriteBufferFromS3.cpp:225]: Will do it

File: src/IO/WriteBufferFromS3.cpp [hunk 11]
```diff
@@ -40,19 +42,135 @@ namespace ProfileEvents
...
+                         std::ios_base::openmode mode) override
+        {
+            return seekoff(off_type(sp), std::ios_base::beg, mode);
+        }
+
+        char * begin = nullptr;
+        size_t size = 0;
+};
+
+struct MemoryStream: std::iostream
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/IO/WriteBufferFromS3.cpp:110]: Is it possible to use something existing? You might have made such classes already. If not, I will separate this class as you suggest.

**Comment by CheSema** [comment]
I would like to discuss the whole approach.
This PR has the most values in the test set. The one-copy optimization is minor.

I see that there are tests which measure the memory consuming. And this PR make things bad for test. For each file writer allocates 33M now.

What is possible to do.
1 - make custom setting and override max_single_part_upload_size, make it smaller
2 - make the reallocation for the first part. I mean that we allocate 1Mb first. If it is not enough, reallocate to x2 and so on untill size hits max_single_part_upload_size. When it does, first part is ready to upload, writer detaches it in order to decide would it be put single part or multi upload.

As for me
1 - is a hack.
2 - is better.

File: src/IO/S3/Client.cpp (hunk 7 above)
Comment (vitlibar) [reply src/IO/S3/Client.cpp:186]: I understand they're different functions, but it looks like `S3::Client::ListObjectsV2()` used to call `Aws::S3::S3Client::ListObjectsV2()`, and now with your changes `S3::Client::ListObjectsV2()` calls `S3::Client::ListObjectsV2()`.

Comment (vitlibar) [reply src/IO/S3/Client.cpp:186]: @antonio2368

Comment (antonio2368) [reply src/IO/S3/Client.cpp:186]: yeah it isn't very clear at first, but to me, the functions are simple enough and maybe add the comment above the first implementation (`ListObjectsV2`) something like `For each request, we wrap the request functions from Aws::S3::Client with doRequest`

File: src/IO/WriteBufferFromS3.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/IO/WriteBufferFromS3.cpp:110]: I don't think we have something existing for that.

File: src/IO/WriteBufferFromFileDecorator.cpp (hunk 9 above)
Comment (vitlibar) [reply src/IO/WriteBufferFromFileDecorator.cpp:42]: Please add a comment saying that.

File: src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp (hunk 8 above)
Comment (vitlibar) [reply src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp:165]: Maybe we can make `finalize()` call `sync()` for files?

**Comment by vitlibar** [comment]
Let's rename this PR? It's not about tests anymore, it's about detaching buffers to eliminate unnecessary copy, isn't it?

File: src/IO/S3/Client.cpp (hunk 7 above)
Comment (vitlibar) [reply src/IO/S3/Client.cpp:186]: I don't feel good about this thing. Maybe we can at least add some comments?

**Comment by vitlibar** [comment]
> What is possible to do. 1 - make custom setting and override max_single_part_upload_size, make it smaller 2 - make the reallocation for the first part. I mean that we allocate 1Mb first. If it is not enough, reallocate to x2 and so on untill size hits max_single_part_upload_size. When it does, first part is ready to upload, writer detaches it in order to decide would it be put single part or multi upload.

Let's do reallocation, `Memory` has the function `resize()` so it must not be difficult.

**Comment by vitlibar** [comment]
> This PR has the most values in the test set. The one-copy optimization is minor.

For me tests are necessary but always less important thing than the code itself.

File: src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp (hunk 8 above)
Comment (CheSema) [reply src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp:165]: I'm not sure that it is good idea. For files sync is omitted usually. Sync is called only if it is strictly necessary or under a settings/flag. This is done due to performance.

File: src/IO/WriteBufferFromFileDecorator.cpp (hunk 9 above)
Comment (CheSema) [reply src/IO/WriteBufferFromFileDecorator.cpp:45]: Done

File: src/IO/S3/Client.cpp (hunk 7 above)
Comment (CheSema) [reply src/IO/S3/Client.cpp:257]: Made some comment here

**Comment by CheSema** [comment]
Noticed a bug in master
https://github.com/ClickHouse/ClickHouse/blob/master/src/IO/WriteBufferFromS3.cpp#L496
What do we do when max_retry exhausted? 
Will open separate PR to fix it ASAP

File: src/IO/WriteBufferFromS3TaskTracker.h [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    bool isAsync() const;
+    void getReady();
+    void getAll();
+    void safeWaitAll();
+    void add(Callback && func);
+
+private:
+    bool is_async;
+    ThreadPoolCallbackRunner<void> scheduler;
+    std::list<std::future<void>> futures;
```
Comment (KochetovNicolai) [inline src/IO/WriteBufferFromS3TaskTracker.h:31]: The think I don't like about std::future is that it contains [exception, mutex and condvar](https://github.com/ClickHouse/libcxx/blob/172b2ae074f6755145b91c53a95c8540c1468239/include/future#L536-L539). Creating it for every task may be a bit overkill (but probably ok for this case).

Comment (CheSema) [reply src/IO/WriteBufferFromS3TaskTracker.h:31]: I share your opinion. But it is created anyway by ThreadPoolCallbackRunner<void>. Packaged task stores promise and correctly wraps all the exceptions to the future.
As I understand. All that staff is stored in the associated state which is shared between future and promise. That state lives while at least either promise or future lives. The promise is preserved as long as function waits execution and execution. Therefore I just not throw all that staff and use for good.

Comment (KochetovNicolai) [reply src/IO/WriteBufferFromS3TaskTracker.h:31]: Indeed ... :(

**Comment by CheSema** [comment]
Releasing resources before detaching
https://github.com/ClickHouse/ClickHouse/pull/47098

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 7d146e3462eed313330be3107000bd2ed75d72ad with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/44869/7d146e3462eed313330be3107000bd2ed75d72ad/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by CheSema** [comment]
Integration Tests (release)
test_part_moves_between_shards/test.py::test_deduplication_while_move flaky
https://github.com/ClickHouse/ClickHouse/issues/49574

**Comment by CheSema** [comment]
ClickHouse Stateless Tests (asan) 02723_zookeeper_name
https://github.com/ClickHouse/ClickHouse/pull/49592

**Comment by CheSema** [comment]
Stateless Tests (aarch64) 
01710_projection_aggregation_in_order
https://github.com/ClickHouse/ClickHouse/issues/49552

**Comment by CheSema** [comment]
Stateless Tests (aarch64)
01710_projection_aggregation_in_order
https://github.com/ClickHouse/ClickHouse/issues/49552

**Comment by CheSema** [comment]
Stress Test (ubsan)
https://s3.amazonaws.com/clickhouse-test-reports/44869/22f7aa8d89107910f10c8ff5fb92296856c81283/stress_test__ubsan_.html
https://github.com/ClickHouse/ClickHouse/issues/49793

File: src/IO/WriteBuffer.h [hunk 13]
```diff
@@ -42,7 +42,8 @@ class WriteBuffer : public BufferBase
     {
         if (!offset())
             return;
-        bytes += offset();
```
Comment (vitlibar) [inline src/IO/WriteBuffer.h:45]: `bytes += offset()` seems more clear

File: src/IO/WriteBufferFromS3MemoryStream.h [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include <iostream>
+
+namespace DB
+{
+
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3MemoryStream.h:11]: Why is the file is named `WriteBufferFromS3MemoryStream`? And please add a comment here for this class.

File: src/IO/WriteBufferFromS3MemoryStream.h [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+struct MemoryStream: std::iostream
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3MemoryStream.h:12]: `std::iostream` is a class, so I suggest making MemoryStream a class too. And usually we use public inheritance.

Comment (vitlibar) [inline src/IO/WriteBufferFromS3MemoryStream.h:12]: Let's rename `MemoryStream` -> `StdStreamFromMemory` ? Because the whole idea of it is to make a `std::iostream`-compatible wrapper around a block in memory.

File: src/IO/WriteBufferFromS3MemoryStream.h [hunk 16]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3MemoryStream.h:5]: Not necessary for this class.

File: src/IO/WriteBufferFromS3BufferAllocationPolicy.h [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include <Storages/StorageS3Settings.h>
+
+namespace DB
+{
+
+struct IBufferAllocationPolicy
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3BufferAllocationPolicy.h:12]: This class is related only to internal stuff inside WriteBufferFromS3, so it seems like it should be declared inside `WriteBufferFromS3`.

File: src/IO/WriteBufferFromS3TaskTracker.h [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include "WriteBufferFromS3.h"
+
+namespace DB
+{
+
+class WriteBufferFromS3::TaskTracker
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3TaskTracker.h:12]: Could we move this class to `WriteBufferFromS3.cpp`? Because this class doesn't seem to be needed in a header.

File: src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp (hunk 8 above)
Comment (vitlibar) [reply src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp:165]: `finalize()` is usually the last thing we do with any write buffer. Could we call `finalize()` after `sync()`?

File: src/IO/WriteBuffer.h (hunk 13 above)
Comment (CheSema) [reply src/IO/WriteBuffer.h:45]: I have noticed in that task some minor bug.

Here is an explanation:
https://github.com/ClickHouse/clickhouse-private/issues/2767#issuecomment-1404291410

> It leads to incorrect count() inside nextImpl(). There offset() is count twice. But I have not found any significant cases when count is called inside nextImpl(). Only for pregress bar in ReadWriteBufferFromHTTPBase for next_callback, it is progress tracker as I see.
For readBuffer problem the same.

May be you wish a comment here instead reversing the code?

Comment (vitlibar) [reply src/IO/WriteBuffer.h:45]: I don't understand how it could cause offset() to be counted twice. Can you explain that a little bit?

File: src/IO/WriteBufferFromS3TaskTracker.h [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    using Callback = std::function<void()>;
+
+    explicit TaskTracker(ThreadPoolCallbackRunner<void> scheduler_);
+    ~TaskTracker();
+
+    static ThreadPoolCallbackRunner<void> syncRunner();
+
+    bool isAsync() const;
+    void getReady();
+    void getAll();
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3TaskTracker.h:24]: Why doesn't the function `getAll()` return anything. It's called `get all` so probably it should return all of something. Or maybe it's just a comment needed here.

File: src/IO/WriteBufferFromFileDecorator.cpp (hunk 6 above)
Comment (vitlibar) [reply src/IO/WriteBufferFromFileDecorator.cpp:27]: Maybe it's just lack of comments. But it's still unclear to me why do we need so many operations. Because the code

```
SwapHelper swap(*this, *impl);
impl->finalize();
```

means something like

```
tmp1 = this->working_buffer;
tmp2 = this->internal_buffer;
tmp3 = this->pos;
this->working_buffer = impl->working_buffer;
this->internal_buffer = impl->internal_buffer;
this->pos = impl->pos;
impl->working_buffer = tmp1;
impl->internal_buffer = tmp2;
impl->pos = tmp3;

try
{
    impl->finalize();

    tmp1 = this->working_buffer;
    tmp2 = this->internal_buffer;
    tmp3 = this->pos;
    this->working_buffer = impl->working_buffer;
    this->internal_buffer = impl->internal_buffer;
    this->pos = impl->pos;
    impl->working_buffer = tmp1;
    impl->internal_buffer = tmp2;
    impl->pos = tmp3;
}
catch (...)
{
    tmp1 = this->working_buffer;
    tmp2 = this->internal_buffer;
    tmp3 = this->pos;
    this->working_buffer = impl->working_buffer;
    this->internal_buffer = impl->internal_buffer;
    this->pos = impl->pos;
    impl->working_buffer = tmp1;
    impl->internal_buffer = tmp2;
    impl->pos = tmp3;

    throw;
}
```

File: src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp [hunk 20]
```diff
@@ -225,10 +231,6 @@ void CachedOnDiskWriteBufferFromFile::nextImpl()
 
         throw;
     }
-
-    /// Write data to cache.
-    cacheData(working_buffer.begin(), size, throw_on_error_from_cache);
-    current_download_offset += size;
```
Comment (vitlibar) [inline src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp:231]: why was this changed?

File: src/IO/WriteBufferFromS3TaskTracker.h (hunk 18 above)
Comment (vitlibar) [inline src/IO/WriteBufferFromS3TaskTracker.h:12]: Please add a comment for this class.

File: src/IO/WriteBufferFromS3BufferAllocationPolicy.h [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+{
+    virtual size_t getNumber() const = 0;
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3BufferAllocationPolicy.h:14]: Please rename this function or add a comment. `getNumber()` looks too general.

**Review by vitlibar — APPROVED** [review]

File: src/IO/WriteBuffer.h (hunk 13 above)
Comment (CheSema) [reply src/IO/WriteBuffer.h:45]: Yes, sure :)

Look at this: https://github.com/ClickHouse/ClickHouse/blob/212aef3dc21ce2b30b0a86b8e67ce7a0818bfebf/src/IO/BufferBase.h#L92

And after at this:
https://github.com/ClickHouse/ClickHouse/blob/212aef3dc21ce2b30b0a86b8e67ce7a0818bfebf/src/IO/WriteBuffer.h#L45-L50

If you call count() inside nextImpl() it add offset() twice. I step into it and notice that. By a lucky chance CH doesn't call count inside nextImpl except minor cases like that:
https://github.com/ClickHouse/ClickHouse/blob/212aef3dc21ce2b30b0a86b8e67ce7a0818bfebf/src/IO/ReadWriteBufferFromHTTP.cpp#L443

File: src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp (hunk 20 above)
Comment (CheSema) [reply src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp:231]: This is test proved change. Actually the is just a bug which has not been triggered before but triggered now.

CachedOnDiskWriteBufferFromFile writes the data from working_buffer to the cache. It is exactly the buffer WriteBufferFromS3 allocates and maintains. The is no any guranty that the data still remain in the working_buffer after next is called. Just none. So the only right way to read from working_buffer in CachedOnDiskWriteBufferFromFile is to read from it before then working_buffer is passed to WriteBufferFromS3 by calling next. Ksenia and I discussed that changes in person.

File: src/IO/StdIStreamFromMemory.h [hunk 22]
```diff
@@ -1,15 +1,15 @@
...
 
 namespace DB
 {
 
-struct MemoryStream: std::iostream
+/// StdIStreamFromMemory is used in WriteBufferFromS3 as a stream which is passed to the S3::Client
+/// It provides istream interface (only reading) over the memory.
+/// However S3::Client requires iostream interface it only reads from the stream
+
+class StdIStreamFromMemory : public std::iostream
```
Comment (vitlibar) [inline src/IO/StdIStreamFromMemory.h:12]: This class is a subclass of `std::iostream`, not `std::istream`.

Comment (CheSema) [reply src/IO/StdIStreamFromMemory.h:12]: Yes, S3::Client is requires std::iostream interface. but writing to it is prohibited.

Comment (vitlibar) [reply src/IO/StdIStreamFromMemory.h:12]: Ok. Apparently it's bad design of AWS that they require `std::iostream` while they only need `std::istream`.
